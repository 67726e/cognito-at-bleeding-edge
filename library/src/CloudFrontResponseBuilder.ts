import { CloudFrontRequestResult } from 'aws-lambda';

export interface CloudFrontResponseBuilder {
  buildCognitoRedirectResponse(location: string): CloudFrontRequestResult;
  buildCookieRedirectResponse(location: string, cookies: string[]): CloudFrontRequestResult;
}

export class DefaultCloudFrontResponseBuilder implements CloudFrontResponseBuilder {
  buildCognitoRedirectResponse(location: string): CloudFrontRequestResult {
    return {
      status: '302',
      headers: {
        'location': [{
          key: 'Location',
          value: location,
        }],
        'cache-control': [{
          key: 'Cache-Control',
          value: 'no-cache, no-store, max-age=0, must-revalidate',
        }],
        'pragma': [{
          key: 'Pragma',
          value: 'no-cache',
        }],
      },
    };
  }

  buildCookieRedirectResponse(location: string, cookies: string[]): CloudFrontRequestResult {
    return {
      status: '302' ,
      headers: {
        'location': [{
          key: 'Location',
          value: location,
        }],
        'cache-control': [{
          key: 'Cache-Control',
          value: 'no-cache, no-store, max-age=0, must-revalidate',
        }],
        'pragma': [{
          key: 'Pragma',
          value: 'no-cache',
        }],
        'set-cookie': cookies.map((cookie) => ({ key: 'Set-Cookie', value: cookie })),
      },
    };
  }
}
