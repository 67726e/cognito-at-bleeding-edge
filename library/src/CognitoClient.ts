import axios from 'axios';
import { stringify } from 'querystring';

import { Logger } from './Logger';
import { CognitoConfiguration, Tokens } from './types';

export interface CognitoClient {
  postAuthorizationCodeExchange(redirectURI: string, code: string): Promise<Tokens>;
  postRefreshTokenExchange(redirectURI: string, refreshToken: string): Promise<any>;
}

export class DefaultCognitoClient implements CognitoClient {
  constructor(
    private readonly cognitoConfiguration: CognitoConfiguration,
    private readonly logger: Logger,
  ) {}

  /**
   * Exchange Authorization Code for Access, ID, Refresh Tokens
   * @param  {String} redirectURI Redirection URI.
   * @param  {String} code        Authorization Code.
   * @return {Promise} Authorized User Tokens.
   */
  postAuthorizationCodeExchange(redirectURI: string, code: string): Promise<Tokens> {
    // TODO: Document Me! - Why Optional `userPoolAppStream`
    const authorization = this.cognitoConfiguration.userPoolAppSecret && Buffer.from(`${this.cognitoConfiguration.userPoolAppId}:${this.cognitoConfiguration.userPoolAppSecret}`).toString('base64');

    const request = {
      url: `https://${this.cognitoConfiguration.userPoolDomain}/oauth2/token`,
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        ...(authorization && {'Authorization': `Basic ${authorization}`}),
      },
      data: stringify({
        client_id:	this.cognitoConfiguration.userPoolAppId,
        code:	code,
        grant_type:	'authorization_code',
        redirect_uri:	redirectURI,
      }),
    } as const;

    this.logger.debug({ msg: 'Fetching Tokens for Authorization Code', request, code });

    return axios.request(request)
      .then((response) => {
        this.logger.debug({ msg: 'Successfully Fetched Tokens for Authorization Code', tokens: response.data });

        // TODO: Validate / Assert Type!
        return {
          accessToken: response.data.access_token,
          idToken: response.data.id_token,
          refreshToken: response.data.refresh_token,
        };
      })
      .catch((error) => {
        this.logger.error({ msg: 'Failure Fetching Tokens for Authorization Code', request, code });

        throw error;
      });
  }

  /**
   * Exchange Refresh Token for Access, ID Tokens
   * @param  {String} redirectURI Redirection URI.
   * @param  {String} refreshToken Refresh Token.
   * @return {Promise<Tokens>} Refreshed User Tokens.
   */
  postRefreshTokenExchange(redirectURI: string, refreshToken: string) {
    // TODO: Document Me! - Why Optional `userPoolAppStream`
    const authorization = this.cognitoConfiguration.userPoolAppSecret && Buffer.from(`${this.cognitoConfiguration.userPoolAppId}:${this.cognitoConfiguration.userPoolAppSecret}`).toString('base64');

    const request = {
      url: `https://${this.cognitoConfiguration.userPoolDomain}/oauth2/token`,
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        ...(authorization && {'Authorization': `Basic ${authorization}`}),
      },
      data: stringify({
        client_id: this.cognitoConfiguration.userPoolAppId,
        refresh_token: refreshToken,
        grant_type: 'refresh_token',
        redirect_uri: redirectURI,
      }),
    } as const;

    this.logger.debug({ msg: 'Fetching Tokens for Refresh Token', request, refreshToken });

    return axios.request(request)
      .then((response) => {
        this.logger.debug({ msg: 'Successfully Fetched Tokens for Refresh Token', tokens: response.data });
        
        // TODO: Validate / Assert Type!
        return {
          accessToken: response.data.access_token,
          idToken: response.data.id_token,
          // TODO: Add `refreshToken` ???
        };
      })
      .catch((error) => {
        this.logger.error({ msg: 'Failure Fetching Tokens for Refresh Token', request, refreshToken });
        
        throw error;
      });
  }
}
