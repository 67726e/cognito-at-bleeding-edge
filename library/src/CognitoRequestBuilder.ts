import { CognitoConfiguration } from './types';

export interface CognitoRequestBuilder {
  buildAuthorizeRequest(redirectURI: string, redirectPath: string): string;
}

export class DefaultCognitoRequestBuilder implements CognitoRequestBuilder {
  constructor(
    private readonly cognitoConfiguration: CognitoConfiguration,
  ) {}

  buildAuthorizeRequest(redirectURI: string, redirectPath: string): string {
    const authorizeUrl = `https://${this.cognitoConfiguration.userPoolDomain}/authorize?` +
      `client_id=${this.cognitoConfiguration.userPoolAppId}&` + 
      `redirect_uri=${redirectURI}&` +
      'response_type=code&' +
      `state=${redirectPath}`;

    return authorizeUrl;
  }
}
