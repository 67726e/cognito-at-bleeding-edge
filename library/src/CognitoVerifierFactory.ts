import { CognitoJwtVerifier, CognitoJwtVerifierProperties, CognitoJwtVerifierSingleUserPool } from 'aws-jwt-verify/cognito-verifier';
import { CognitoConfiguration } from './types';

export interface CognitoTokenVerifierFactory {
  createAccessTokenVerifier(): CognitoJwtVerifierSingleUserPool<CognitoJwtVerifierProperties>;
  createIdTokenVerifier(): CognitoJwtVerifierSingleUserPool<CognitoJwtVerifierProperties>;
}

export class DefaultCognitoTokenVerifierFactory implements CognitoTokenVerifierFactory {
  constructor(
    private readonly cognitoConfiguration: CognitoConfiguration,
  ) {}

  createAccessTokenVerifier() {
    return CognitoJwtVerifier.create({
      userPoolId: this.cognitoConfiguration.userPoolId,
      clientId: this.cognitoConfiguration.userPoolAppId,
      tokenUse: 'access',
    } as CognitoJwtVerifierProperties);
  }

  createIdTokenVerifier() {
    return CognitoJwtVerifier.create({
      userPoolId: this.cognitoConfiguration.userPoolId,
      clientId: this.cognitoConfiguration.userPoolAppId,
      tokenUse: 'id',
    } as CognitoJwtVerifierProperties);
  }
}
