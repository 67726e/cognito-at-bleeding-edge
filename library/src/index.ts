export * from './Authenticator';
export * from './CloudFrontResponseBuilder';
export * from './CognitoClient';
export * from './CognitoRequestBuilder';
export * from './CognitoVerifierFactory';
export * from './cookie-at-bleeding-edge';
export * from './CookieRepository';
export * from './DefaultAuthenticator';
export * from './Logger';
export * from './types';