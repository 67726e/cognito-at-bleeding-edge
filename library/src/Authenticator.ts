import { CloudFrontRequest, CloudFrontRequestEvent, CloudFrontRequestResult } from 'aws-lambda';
import { parse } from 'querystring';

import { CloudFrontResponseBuilder, } from './CloudFrontResponseBuilder';
import { CognitoClient, } from './CognitoClient';
import { CognitoRequestBuilder, } from './CognitoRequestBuilder';
import { SameSite, SAME_SITE_VALUES } from './cookie-at-bleeding-edge';
import { CookieRepository, } from './CookieRepository';
import { Logger } from './Logger';
import { CognitoConfiguration, CookieHeaders, Tokens } from './types';

export type DefaultAuthenticatorCognitoConfiguration = CognitoConfiguration

export interface DefaultAuthenticatorCookieConfiguration {
  disableDomain?: boolean;
  expirationDays?: number;
  httpOnly?: boolean;
  sameSite?: SameSite;
  path?: string;
}

export interface DefaultAuthenticatorConfiguration {
  cognitoConfiguration: DefaultAuthenticatorCognitoConfiguration;
  cookieConfiguration?: DefaultAuthenticatorCookieConfiguration;
}

export type AuthenticatorResult = {
  actual: CloudFrontRequestResult,
  isAuthenticated: boolean,
};

export class Authenticator {
  constructor(
    private readonly cloudFrontResponseBuilder: CloudFrontResponseBuilder,
    private readonly cognitoClient: CognitoClient,
    private readonly cognitoRequestBuilder: CognitoRequestBuilder,
    private readonly cookieRepository: CookieRepository,
    private readonly logger: Logger,
    private readonly accessTokenVerifier: any,
    private readonly idTokenVerifier: any,
  ) {}

  /**
   * Extract value of the authentication token from the request cookies.
   * @param  {Array}  cookieHeaders 'Cookie' request headers.
   * @return {Tokens} Extracted id token or access token. Null if not found.
   */
  _getTokensFromCookie(cookieHeaders: CookieHeaders): Tokens {
    if (!cookieHeaders) {
      this.logger.debug("Cookies weren't present in the request");

      throw new Error("Cookies weren't present in the request");
    }

    this.logger.debug({ msg: 'Extracting authentication token from request cookie', cookieHeaders });

    const tokens: Tokens = this.cookieRepository.parseCookies(cookieHeaders);

    if (!tokens.accessToken && !tokens.idToken && !tokens.refreshToken) {
      this.logger.debug('Neither accessToken, idToken, nor refreshToken was present in request cookies');

      throw new Error('Neither accessToken, idToken, nor refreshToken was present in request cookies');
    }

    this.logger.debug({ msg: 'Found tokens in cookie', tokens });

    return tokens;
  }

  /**
   * Create a Lambda@Edge redirection response to set the tokens on the user's browser cookies.
   * @param  {Object} tokens   Cognito User Pool tokens.
   * @param  {String} domain   Website domain.
   * @param  {String} location Path to redirection.
   * @return Lambda@Edge response.
   */
  _createSetCookieRedirectResponse(tokens: Tokens, domain: string, location: string): CloudFrontRequestResult {
    const cookies = this.cookieRepository.createCookies(domain, tokens);

    const response: CloudFrontRequestResult = this.cloudFrontResponseBuilder.buildCookieRedirectResponse(location, cookies);

    this.logger.debug({ msg: 'Generated set-cookie response', response });

    return response;
  }

  /**
   * Get redirect to cognito userpool response
   * @param  {CloudFrontRequest}  request The original request
   * @param  {string}  redirectURI Redirection URI.
   * @return {CloudFrontRequestResult} Redirect response.
   */
  _createCognitoAuthorizationRedirectResponse(request: CloudFrontRequest, redirectURI: string): CloudFrontRequestResult {
    const redirectPath = (request.querystring && request.querystring !== '') ?
      request.uri + encodeURIComponent('?' + request.querystring) :
      request.uri;

    const userPoolUrl = this.cognitoRequestBuilder.buildAuthorizeRequest(redirectURI, redirectPath);

    this.logger.debug(`Redirecting user to Cognito User Pool URL ${userPoolUrl}`);

    return this.cloudFrontResponseBuilder.buildCognitoRedirectResponse(userPoolUrl);
  }

  /**
   * Handle Lambda@Edge event:
   *   * if authentication cookie is present and valid: forward the request
   *   * if authentication cookie is invalid, but refresh token is present: set cookies with refreshed tokens
   *   * if ?code=<grant code> is present: set cookies with new tokens
   *   * else redirect to the Cognito UserPool to authenticate the user
   * @param  {Object}  event Lambda@Edge event.
   * @return {Promise} CloudFront response.
   */
  async handle(event: CloudFrontRequestEvent): Promise<AuthenticatorResult> {
    this.logger.debug({ msg: 'Handling Lambda@Edge Event', event });

    const { request } = event.Records[0].cf;
    const requestParams = parse(request.querystring);
    const cloudFrontDomain = request.headers.host[0].value;
    const redirectURI = `https://${cloudFrontDomain}`;

    // 1. Get Token(s) from Cookie Header(s)
    //  Cookie(s) May or May Not Include: Access Token, ID Token, Refresh Token
    // 2. Verify Access Token, ID Token
    //  If Access Token and ID Token verify, return original request.
    //  If Access Token and or ID Token fail, next step 
    // 3. Fetch Token(s) from Refresh Token
    //  If Fetch Token(s) succeed, return redirect request w/ Access Token, ID Token as `Set-Cookie` Header
    //  If Fetch Token(s) fail, next step
    // 4. Get Authentication Code from Query Parameter(s), Exchange for Token(s)
    //  If Fetch Token(s) success, return redirect request w/ Access Token, ID Token as `Set-Cookie` Header
    //  If Fetch Token(s) fail, final failure
    // 5. Return Redirect to Cognito Authentication Pool

    try {
      // 1. Get Token(s) from Cookie Header(s)
      const tokens = this._getTokensFromCookie(request.headers.cookie);

      this.logger.debug({ msg: 'Verifying token...', tokens });

      try {
        // 2. Verify Access Token, ID Token
        const identity = await this.idTokenVerifier.verify(tokens.idToken);
        const access = await this.accessTokenVerifier.verify(tokens.accessToken);

        this.logger.info({ msg: 'Forwarding request', path: request.uri, identity, access });

        return { actual: request, isAuthenticated: true, };
      } catch (err) {
        // 3. Fetch Token(s) from Refresh Token
        if (tokens.refreshToken) {
          this.logger.debug({ msg: 'Verifying accessToken, idToken failed, verifying refresh token instead...', tokens, err });

          return await this.cognitoClient.postRefreshTokenExchange(redirectURI, tokens.refreshToken)
            .then(tokens => {
              const result = this._createSetCookieRedirectResponse(tokens, cloudFrontDomain, request.uri);

              return { actual: result, isAuthenticated: false, };
            });
        } else {
          throw err;
        }
      }
    } catch (err) {
      this.logger.debug("User isn't authenticated: %s", err);

      if (requestParams.code) {
        // 4. Get Authentication Code from Query Parameter(s), Exchange for Token(s)
        return await this.cognitoClient.postAuthorizationCodeExchange(redirectURI, requestParams.code as string)
          .then(tokens => {
            const result = this._createSetCookieRedirectResponse(tokens, cloudFrontDomain, requestParams.state as string);

            return { actual: result, isAuthenticated: false, };
          });
      } else {
        // 5. Return Redirect to Cognito Authentication Pool
        const result = this._createCognitoAuthorizationRedirectResponse(request, redirectURI);

        return { actual: result, isAuthenticated: false, };
      }
    }
  }
}
