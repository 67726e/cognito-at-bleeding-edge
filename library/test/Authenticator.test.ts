
import { beforeEach, describe, expect, jest, test } from '@jest/globals'
import { CloudFrontRequestEvent } from 'aws-lambda/trigger/cloudfront-request';

import { Authenticator } from '../src/Authenticator';
import { DefaultCloudFrontResponseBuilder } from '../src/CloudFrontResponseBuilder';
import { CognitoClient, DefaultCognitoClient } from '../src/CognitoClient';
import { DefaultCognitoRequestBuilder } from '../src/CognitoRequestBuilder';
import { CognitoTokenVerifierFactory, } from '../src/CognitoVerifierFactory';
import { DefaultCookieRepository } from '../src/CookieRepository';
import { DefaultAuthenticator } from '../src/DefaultAuthenticator';
import { Logger } from '../src/Logger';
import { CognitoConfiguration, CookieConfiguration, Tokens } from '../src/types';
import { createEventWithCookies, createEventWithoutCookies, createEventWithQuery } from './fixtures/event';



class NOOPLogger implements Logger {
    debug(...args: any): void {}
    error(...args: any): void {}
    info(...args: any): void {}
}

class MockCognitoTokenVerifierFactory implements CognitoTokenVerifierFactory {
    constructor(
        private readonly cognitoConfiguration: CognitoConfiguration,
    ) {}

    // @ts-ignore
    createAccessTokenVerifier(): CognitoJwtVerifierSingleUserPool<CognitoJwtVerifierProperties> {
        return {
            verify: async (accessToken: string) => {
                if (accessToken === 'mockAccessToken') return Promise.resolve();
                else return Promise.reject();
            }
        };
    }

    // @ts-ignore
    createIdTokenVerifier(): CognitoJwtVerifierSingleUserPool<CognitoJwtVerifierProperties> {
        return {
            verify: async (idToken: string) => {
                if (idToken === 'mockIdToken') return Promise.resolve();
                else return Promise.reject();
            }
        };
    }
}

class MockCognitoClient implements CognitoClient {
    constructor(
        private readonly cognitoConfiguration: CognitoConfiguration,
        private readonly logger: Logger,
    ) {}

    postAuthorizationCodeExchange(redirectURI: string, code: string): Promise<Tokens> {
        if (code === 'mockAuthorizationCode') {
            return Promise.resolve({
                accessToken: 'mockAccessToken',
                idToken: 'mockIdToken',
                refreshToken: 'mockRefreshToken',
            });
        } else if (code === 'mockMalformedAuthorizationCode') {
            // Empty Token Response
            return Promise.resolve({});
        } else return Promise.reject();
    }

    postRefreshTokenExchange(redirectURI: string, refreshToken: string): Promise<any> {
        if (refreshToken === 'mockRefreshToken') {
            return Promise.resolve({
                accessToken: 'mockAccessToken',
                idToken: 'mockIdToken',
            });
        } else if (refreshToken === 'mockMalformedRefreshToken') {
            // Empty Token Response
            return Promise.resolve({});
        } else return Promise.reject();
    }
}



describe('Authenticator', () => {
    let authenticator: Authenticator;

    beforeEach(() => {
        jest
            .useFakeTimers()
            .setSystemTime(new Date('2023-10-20'));

        const cognitoConfiguration: CognitoConfiguration = {
            userPoolId: 'us-east-1_abcdef123',
            userPoolAppId: '123456789qwertyuiop987abcd',
            userPoolAppSecret: undefined,
            userPoolDomain: 'mock-cognito-domain.auth.us-east-1.amazoncognito.com',    
        };
        const cookieConfiguration: CookieConfiguration = DefaultAuthenticator.buildCookieConfiguration(undefined);
    
        // NOTE: Mocked to prevent network calls in tests...
        // NOTE: Mocked to simulate behaviors (good token, bad token) with mocked data...
        //  SEE: `MockCognitoTokenVerifierFactory`
        const cognitoTokenVerifierFactory: CognitoTokenVerifierFactory = new MockCognitoTokenVerifierFactory(
          cognitoConfiguration,
        );
    
        // NOTE: Mocked to prevent console output in tests...
        const defaultLogger: Logger = new NOOPLogger();
    
        authenticator = new Authenticator(
          new DefaultCloudFrontResponseBuilder(),
          // NOTE: Mocked to prevent network calls in tests...
          // NOTE: Mocked to simulate behaviors (good request, bad request) with mocked data...
          //    See: `MockCognitoClient`
          new MockCognitoClient(
            cognitoConfiguration,
            defaultLogger,
          ),
          new DefaultCognitoRequestBuilder(
            cognitoConfiguration,
          ),
          new DefaultCookieRepository(
            cookieConfiguration,
          ),
          defaultLogger,
          cognitoTokenVerifierFactory.createAccessTokenVerifier(),
          cognitoTokenVerifierFactory.createIdTokenVerifier(),
        );
    });

    test('handle / missing access tokens', async () => {
        // Missing access token should trigger a redirect to Cognito for authentication.
        const event = createEventWithCookies({ cookie: 'cloudfront.idToken=mockIdToken' });

        const response = await authenticator.handle(event as CloudFrontRequestEvent);

        expect(response).toBeDefined();
        expect(response).toMatchSnapshot();
    });

    test('handle / missing id tokens', async () => {
        // Misisng id token should trigger a redirect to Cognito for authentication.
        const event = createEventWithCookies({ cookie: 'cloudfront.accessToken=mockAccessToken' });

        const response = await authenticator.handle(event as CloudFrontRequestEvent);

        expect(response).toBeDefined();
        expect(response).toMatchSnapshot();
    });

    test('handle / invalid accessToken', async () => {
        // Invalid access token should trigger a redirect to Cognito for authentication.
        const event = createEventWithCookies({ cookie: 'cloudfront.accessToken=invalidAccessToken; cloudfront.idToken=mockIdToken' });

        const response = await authenticator.handle(event as CloudFrontRequestEvent);

        expect(response).toBeDefined();
        expect(response).toMatchSnapshot();
    });

    test('handle / invalid idToken', async () => {
        // Invalid ID token should trigger a redirect to Cognito for authentication.
        const event = createEventWithCookies({ cookie: 'cloudfront.accessToken=mockAccessToken; cloudfront.idToken=invalidIdToken' });

        const response = await authenticator.handle(event as CloudFrontRequestEvent);

        expect(response).toBeDefined();
        expect(response).toMatchSnapshot();
    });

    test('handle / redirect to cognito', async () => {
        // Cookie-less event should trigger a redirect to Cognito for authentication.
        const event = createEventWithoutCookies({});

        const response = await authenticator.handle(event as CloudFrontRequestEvent);

        expect(response).toBeDefined();
        expect(response).toMatchSnapshot();
    });

    test('handle / refresh token exchange', async () => {
        // Missing access token, id token should trigger an redirect with `Set-Cookie` headers
        const event = createEventWithCookies({ cookie: 'cloudfront.refreshToken=mockRefreshToken' });

        const response = await authenticator.handle(event as CloudFrontRequestEvent);

        expect(response).toBeDefined();
        expect(response).toMatchSnapshot();
    });

    test('handle / refresh token exchange, invalid refresh token', async () => {
        // Invalid refresh token should trigger an error
        // Error should trigger a redirect to Cognito for authentication.
        const event = createEventWithCookies({ cookie: 'cloudfront.refreshToken=mockInvalidRefreshToken' });

        const response = await authenticator.handle(event as CloudFrontRequestEvent);

        expect(response).toBeDefined();
        expect(response).toMatchSnapshot();
    });

    // TODO: Review Me!
    // test('handle / refresh token exchange, malformed response', async () => {
    //     // Malformed refresh token should trigger an malformed response from the Cognito mock
    //     //  Malformed response should trigger an error
    //     //  Error should trigger an 
    //     const event = createEventWithCookies({ cookie: 'cloudfront.refreshToken=mockMalformedRefreshToken' });

    //     const response = await authenticator.handle(event as CloudFrontRequestEvent);

    //     expect(response).toBeDefined();
    //     console.log(JSON.stringify(response, null, 4))
    //     // expect(response).toMatchSnapshot();
    // });

    test('handle / authorization code exchange', async () => {
        // Event with authorization code query parameter should trigger an redirect with `Set-Cookie` headers
        const event = createEventWithQuery({ query: 'code=mockAuthorizationCode' });

        const response = await authenticator.handle(event as CloudFrontRequestEvent);

        expect(response).toBeDefined();
        expect(response).toMatchSnapshot();
    });

    // TODO: Review Me!
    // test('handle / authorization code exchange, invalid authorization code', async () => {
    //     // Invalid authorization code should trigger an error
    //     const event = createEventWithQuery({ query: 'code=mockInvalidAuthorizationCode' });

    //     const response = await authenticator.handle(event as CloudFrontRequestEvent);

    //     expect(response).toBeDefined();
    //     expect(response).toMatchSnapshot();
    // });

    // test('handle / authorization code, malformed response', async () => {
    //     // Event with authorization code query parameter should trigger an redirect with `Set-Cookie` headers
    //     const event = createEventWithQuery({ query: 'code=mockMalformedAuthorizationCode' });

    //     const response = await authenticator.handle(event as CloudFrontRequestEvent);

    //     expect(response).toBeDefined();
    //     console.log(JSON.stringify(response, null, 4));
    //     // expect(response).toMatchSnapshot();
    // });

    test('handle / forward to distribution', async () => {
        // Valid access, identity tokens should allow the original request to be forwarded.
        const event = createEventWithCookies({ cookie: 'cloudfront.accessToken=mockAccessToken; cloudfront.idToken=mockIdToken' });

        const response = await authenticator.handle(event as CloudFrontRequestEvent);

        expect(response).toBeDefined();
        expect(response).toMatchSnapshot();
    });
});
