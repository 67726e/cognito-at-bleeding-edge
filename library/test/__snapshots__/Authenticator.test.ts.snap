// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authenticator handle / authorization code exchange 1`] = `
{
  "headers": {
    "cache-control": [
      {
        "key": "Cache-Control",
        "value": "no-cache, no-store, max-age=0, must-revalidate",
      },
    ],
    "location": [
      {
        "key": "Location",
        "value": undefined,
      },
    ],
    "pragma": [
      {
        "key": "Pragma",
        "value": "no-cache",
      },
    ],
    "set-cookie": [
      {
        "key": "Set-Cookie",
        "value": "cloudfront.accessToken=mockAccessToken; Domain=d20o2ten651kje.cloudfront.net; Path=/; Expires=Sat, 19 Oct 2024 00:00:00 GMT; Secure",
      },
      {
        "key": "Set-Cookie",
        "value": "cloudfront.idToken=mockIdToken; Domain=d20o2ten651kje.cloudfront.net; Path=/; Expires=Sat, 19 Oct 2024 00:00:00 GMT; Secure",
      },
      {
        "key": "Set-Cookie",
        "value": "cloudfront.refreshToken=mockRefreshToken; Domain=d20o2ten651kje.cloudfront.net; Path=/; Expires=Sat, 19 Oct 2024 00:00:00 GMT; Secure",
      },
    ],
  },
  "status": "302",
}
`;

exports[`Authenticator handle / forward to distribution 1`] = `
{
  "clientIp": "12.34.567.890",
  "headers": {
    "accept": [
      {
        "key": "accept",
        "value": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8",
      },
    ],
    "accept-encoding": [
      {
        "key": "accept-encoding",
        "value": "gzip, deflate, br",
      },
    ],
    "accept-language": [
      {
        "key": "accept-language",
        "value": "en-US,en;q=0.9",
      },
    ],
    "cache-control": [
      {
        "key": "cache-control",
        "value": "max-age=0",
      },
    ],
    "cookie": [
      {
        "key": "Cookie",
        "value": "cloudfront.accessToken=mockAccessToken; cloudfront.idToken=mockIdToken",
      },
    ],
    "host": [
      {
        "key": "Host",
        "value": "d20o2ten651kje.cloudfront.net",
      },
    ],
    "if-modified-since": [
      {
        "key": "if-modified-since",
        "value": "Fri, 21 Jul 2023 14:52:55 GMT",
      },
    ],
    "if-none-match": [
      {
        "key": "if-none-match",
        "value": ""fdb8756628111e736d7f6b238664ae01"",
      },
    ],
    "sec-fetch-dest": [
      {
        "key": "sec-fetch-dest",
        "value": "document",
      },
    ],
    "sec-fetch-mode": [
      {
        "key": "sec-fetch-mode",
        "value": "navigate",
      },
    ],
    "sec-fetch-site": [
      {
        "key": "sec-fetch-site",
        "value": "none",
      },
    ],
    "sec-fetch-user": [
      {
        "key": "sec-fetch-user",
        "value": "?1",
      },
    ],
    "sec-gpc": [
      {
        "key": "sec-gpc",
        "value": "1",
      },
    ],
    "upgrade-insecure-requests": [
      {
        "key": "upgrade-insecure-requests",
        "value": "1",
      },
    ],
    "user-agent": [
      {
        "key": "User-Agent",
        "value": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/112.0.0.0 Safari/537.36",
      },
    ],
  },
  "method": "GET",
  "querystring": "",
  "uri": "/",
}
`;

exports[`Authenticator handle / invalid accessToken 1`] = `
{
  "headers": {
    "cache-control": [
      {
        "key": "Cache-Control",
        "value": "no-cache, no-store, max-age=0, must-revalidate",
      },
    ],
    "location": [
      {
        "key": "Location",
        "value": "https://mock-cognito-domain.auth.us-east-1.amazoncognito.com/authorize?client_id=123456789qwertyuiop987abcd&redirect_uri=https://d20o2ten651kje.cloudfront.net&response_type=code&state=/",
      },
    ],
    "pragma": [
      {
        "key": "Pragma",
        "value": "no-cache",
      },
    ],
  },
  "status": "302",
}
`;

exports[`Authenticator handle / invalid idToken 1`] = `
{
  "headers": {
    "cache-control": [
      {
        "key": "Cache-Control",
        "value": "no-cache, no-store, max-age=0, must-revalidate",
      },
    ],
    "location": [
      {
        "key": "Location",
        "value": "https://mock-cognito-domain.auth.us-east-1.amazoncognito.com/authorize?client_id=123456789qwertyuiop987abcd&redirect_uri=https://d20o2ten651kje.cloudfront.net&response_type=code&state=/",
      },
    ],
    "pragma": [
      {
        "key": "Pragma",
        "value": "no-cache",
      },
    ],
  },
  "status": "302",
}
`;

exports[`Authenticator handle / missing access tokens 1`] = `
{
  "headers": {
    "cache-control": [
      {
        "key": "Cache-Control",
        "value": "no-cache, no-store, max-age=0, must-revalidate",
      },
    ],
    "location": [
      {
        "key": "Location",
        "value": "https://mock-cognito-domain.auth.us-east-1.amazoncognito.com/authorize?client_id=123456789qwertyuiop987abcd&redirect_uri=https://d20o2ten651kje.cloudfront.net&response_type=code&state=/",
      },
    ],
    "pragma": [
      {
        "key": "Pragma",
        "value": "no-cache",
      },
    ],
  },
  "status": "302",
}
`;

exports[`Authenticator handle / missing id tokens 1`] = `
{
  "headers": {
    "cache-control": [
      {
        "key": "Cache-Control",
        "value": "no-cache, no-store, max-age=0, must-revalidate",
      },
    ],
    "location": [
      {
        "key": "Location",
        "value": "https://mock-cognito-domain.auth.us-east-1.amazoncognito.com/authorize?client_id=123456789qwertyuiop987abcd&redirect_uri=https://d20o2ten651kje.cloudfront.net&response_type=code&state=/",
      },
    ],
    "pragma": [
      {
        "key": "Pragma",
        "value": "no-cache",
      },
    ],
  },
  "status": "302",
}
`;

exports[`Authenticator handle / redirect to cognito 1`] = `
{
  "headers": {
    "cache-control": [
      {
        "key": "Cache-Control",
        "value": "no-cache, no-store, max-age=0, must-revalidate",
      },
    ],
    "location": [
      {
        "key": "Location",
        "value": "https://mock-cognito-domain.auth.us-east-1.amazoncognito.com/authorize?client_id=123456789qwertyuiop987abcd&redirect_uri=https://d20o2ten651kje.cloudfront.net&response_type=code&state=/",
      },
    ],
    "pragma": [
      {
        "key": "Pragma",
        "value": "no-cache",
      },
    ],
  },
  "status": "302",
}
`;

exports[`Authenticator handle / refresh token exchange 1`] = `
{
  "headers": {
    "cache-control": [
      {
        "key": "Cache-Control",
        "value": "no-cache, no-store, max-age=0, must-revalidate",
      },
    ],
    "location": [
      {
        "key": "Location",
        "value": "/",
      },
    ],
    "pragma": [
      {
        "key": "Pragma",
        "value": "no-cache",
      },
    ],
    "set-cookie": [
      {
        "key": "Set-Cookie",
        "value": "cloudfront.accessToken=mockAccessToken; Domain=d20o2ten651kje.cloudfront.net; Path=/; Expires=Sat, 19 Oct 2024 00:00:00 GMT; Secure",
      },
      {
        "key": "Set-Cookie",
        "value": "cloudfront.idToken=mockIdToken; Domain=d20o2ten651kje.cloudfront.net; Path=/; Expires=Sat, 19 Oct 2024 00:00:00 GMT; Secure",
      },
    ],
  },
  "status": "302",
}
`;

exports[`Authenticator handle / refresh token exchange, invalid refresh token 1`] = `
{
  "headers": {
    "cache-control": [
      {
        "key": "Cache-Control",
        "value": "no-cache, no-store, max-age=0, must-revalidate",
      },
    ],
    "location": [
      {
        "key": "Location",
        "value": "https://mock-cognito-domain.auth.us-east-1.amazoncognito.com/authorize?client_id=123456789qwertyuiop987abcd&redirect_uri=https://d20o2ten651kje.cloudfront.net&response_type=code&state=/",
      },
    ],
    "pragma": [
      {
        "key": "Pragma",
        "value": "no-cache",
      },
    ],
  },
  "status": "302",
}
`;
